crawl-ref/source/webserver/game_data/static/consumables.js

```diff
@@ -0,0 +1,134 @@
+        {
+            $canvas.addClass("empty");
+            return;
+        }
+
+        // Filter
+        var filtered_inv = Object.values(player.inv).filter(function(item) {
+            if (!item.quantity
+                || !base_types.includes(item.base_type)
+                || (!show_unidentified && !item.flags)
+                || filters.some(function(re) { return item.name.match(re); }))
+            {
+                return false;
+            }
+            else
+                return true;
+        });
+
+        if (!filtered_inv.length)
+        {
+            $canvas.addClass("empty");
+            return;
+        }
+
+        // Sort
+        filtered_inv.sort(function(a, b) {
+            if (a.base_type === b.base_type)
+                return a.sub_type - b.sub_type;
+
+            return a.base_type - b.base_type;
+        });
+
+        // Render
+        var required_width = renderer.cell_width * filtered_inv.length * scale;
+        var available_width = $("#dungeon").width() - borders_width;
+        var max_cells = Math.floor(available_width / (renderer.cell_width * scale));
+        var panel_width = Math.min(required_width, available_width);
+
+        util.init_canvas($canvas[0], panel_width, renderer.cell_height * scale);
+        renderer.init($canvas[0]);
+
+        renderer.ctx.fillStyle = "black";
+        renderer.ctx.fillRect(0, 0, panel_width, renderer.cell_height * scale);
+
+        filtered_inv.slice(0, max_cells).forEach(function(i, idx) {
+            i.tile.forEach(function(t) {
+                renderer.draw_main(t, renderer.cell_width * idx * scale, 0, scale);
+            });
+
+            var qty;
+
+            if (i.base_type === enums.base_type.OBJ_MISCELLANY
+                && i.sub_type !== enums.misc_item_type.MISC_ZIGGURAT)
+            {
+                qty = i.plus;
+            }
+            else
+                qty = i.plus || i.quantity;
+
+            renderer.draw_quantity(qty, renderer.cell_width * idx * scale, 0, scale);
+        });
+
+        if (available_width < required_width)
+        {
+            var ellipsis = icons.ELLIPSIS;
+            var x_pos = available_width - icons.get_tile_info(ellipsis).w * scale;
+            renderer.draw_icon(ellipsis, x_pos, 0, -2, -2, scale);
+        }
+        $canvas.removeClass("empty");
+    }
+
+    options.add_listener(function () {
+        var update_required = false;
+
+        var new_base_types = options.get("consumables");
+        if (base_types.toString() !== new_base_types.toString())
+        {
+            base_types = new_base_types;
+            update_required = true;
+        }
+
+        var new_filters = options.get("consumables_filter");
+        if (filters_text !== new_filters.toString())
+        {
+            filters_text = new_filters.toString();
+
+            filters = [];
+            new_filters.forEach(function(pattern) {
+                filters.push(new RegExp(pattern));
```

Comment:
This would be a big change, so I wouldn't necessarily want to hold up the PR with it, but it might be preferable to implement the filtering on the crawl binary side. Benefits: (i) use the same regex parser that the crawl binary uses (ii) possibly faster (though this would be tbd), and (iii) makes this more easily portable to other build targets. Is this something you have looked into already and could estimate feasibility?

Reply:
Initially I wanted to use a vector of `text_pattern`s for the filter, so I could verify regexps via `text_pattern:valid()`. But it turns out, the game does no such checking for other similar options.
(i) The same regex parser would guarantee the same syntax, but there doesn't seem to be a big difference between crawl and JavaScript regexps. The same basic patterns work fine on both sides:

```
consumables_panel_filter +=  scroll.*(enchant|brand) weapon
autopickup_exceptions    ^= >scroll.*(enchant|brand) weapon
```

(ii) I played a couple of games with the panel on a laptop, which needs half an hour to compile the game, and noticed no slowdown. The worst case would be matching 52 item names against a few dozen regexps each time the player uses or picks up an item. Compared with matching each new game message against hundreds of `force_more_message`s, this shouldn't be too bad. Browsers are slow but not that slow!

(iii) There is a benefit of having filters on the client side. Several existing Web Tiles options can be changed via `window.set_option()` without leaving the game: fonts, `tile_display_mode`, or even minimap colours. Four new options can be changed via the browser console too, which was useful during development.

To summarize, the current state of inventory filtering is good enough, but I agree that filtering should be moved to the crawl binary side. I'll look into that.

Reply:
Done. As a side benefit, there's no need to have (and keep in sync!) `base_type` and `misc_item_type` enums on the JavaScript side.